src/TurbulenceClosures/turbulence_closure_implementations/CATKEVerticalDiffusivities/time_step_turbulent_kinetic_energy.jl

```diff
@@ -0,0 +1,206 @@
+    on_bottom = !inactive_cell(i, j, k, grid) & inactive_cell(i, j, k-1, grid)
+    Δz = Δzᶜᶜᶜ(i, j, k, grid)
+    Cᵂϵ = closure_ij.turbulent_kinetic_energy_equation.Cᵂϵ
+    e⁺ = clip(eⁱʲᵏ)
+    w★ = sqrt(e⁺)
+    div_Jᵉ_e = - on_bottom * Cᵂϵ * w★ / Δz
+
+    # Implicit TKE dissipation
+    ω = dissipation_rate(i, j, k, grid, closure_ij, next_velocities, tracers, buoyancy, diffusivities)
+
+    # The interior contributions to the linear implicit term `L` are defined via
+    #
+    #       ∂t e = Lⁱ e + ⋯,
+    #
+    # So
+    #
+    #       Lⁱ e = wb - ϵ
+    #            = (wb / e - ω) e,
+    #               ↖--------↗
+    #                  = Lⁱ
+    #
+    # where ω = ϵ / e ∼ √e / ℓ.
+
+    @inbounds Le[i, j, k] = wb⁻_e - ω + div_Jᵉ_e
+
+    # Compute fast TKE RHS
+    u⁺ = next_velocities.u
+    v⁺ = next_velocities.v
+    uⁿ = previous_velocities.u
+    vⁿ = previous_velocities.v
+    κu = diffusivities.κu
+
+    # TODO: correctly handle closure / diffusivity tuples
+    # TODO: the shear_production is actually a slow term so we _could_ precompute.
+    P = shear_production(i, j, k, grid, κu, uⁿ, u⁺, vⁿ, v⁺)
+    ϵ = dissipation(i, j, k, grid, closure_ij, next_velocities, tracers, buoyancy, diffusivities)
+    fast_Gⁿe = P + wb⁺ - ϵ
+
+    # Advance TKE and store tendency
+    FT = eltype(χ)
+    Δτ = convert(FT, Δτ)
+
+    # See below.
+    α = convert(FT, 1.5) + χ
+    β = convert(FT, 0.5) + χ
+
+    @inbounds begin
+        total_Gⁿe = slow_Gⁿe[i, j, k] + fast_Gⁿe
+        e[i, j, k] += Δτ * (α * total_Gⁿe - β * G⁻e[i, j, k])
+        G⁻e[i, j, k] = total_Gⁿe
+    end
+end
+
+@inline function implicit_linear_coefficient(i, j, k, grid, closure::FlavorOfCATKE{<:VITD}, K, ::Val{id}, args...) where id
+    L = K._tupled_implicit_linear_coefficients[id]
+    return @inbounds L[i, j, k]
+end
+
+#=
+using Oceananigans.TurbulenceClosures.CATKEVerticalDiffusivities: FlavorOfCATKE
+
+@inline tracer_tendency_kernel_function(model::HFSM, name, c, K)                     = compute_hydrostatic_free_surface_Gc!, c, K
+@inline tracer_tendency_kernel_function(model::HFSM, ::Val{:e}, c::FlavorOfCATKE, K) = compute_hydrostatic_free_surface_Ge!, c, K
+
+function tracer_tendency_kernel_function(model::HFSM, ::Val{:e}, closures::Tuple, diffusivity_fields::Tuple)
+    catke_index = findfirst(c -> c isa FlavorOfCATKE, closures)
+
+    if isnothing(catke_index)
+        return compute_hydrostatic_free_surface_Gc!, closures, diffusivity_fields
+    else
+        catke_closure = closures[catke_index]
+        catke_diffusivity_fields = diffusivity_fields[catke_index]
+        return compute_hydrostatic_free_surface_Ge!, catke_closure, catke_diffusivity_fields 
+    end
+end
+
+@inline function top_tracer_boundary_conditions(grid, tracers)
+    names = propertynames(tracers)
+    values = Tuple(tracers[c].boundary_conditions.top for c in names)
+
+    # Some shenanigans for type stability?
+    return NamedTuple{tuple(names...)}(tuple(values...))
+end
+=#
```

Comment:
do we need these?

Reply:
ah these are for tuples

Reply:
we _will_ need them if we want to run with closures additional to CATKE. If we run only with CATKE we don't need them. I think we should probably keep this code because we want to restore this capability. It's just pointless to support until we can actually verify that CATKE is ready to go in a 3D setting